x-pack/plugins/fleet/server/services/epm/archive/cache.ts

```diff
@@ -3,21 +3,57 @@
  * or more contributor license agreements. Licensed under the Elastic License;
  * you may not use this file except in compliance with the Elastic License.
  */
-import { pkgToPkgKey } from '../registry/index';
+import { ArchiveEntry } from './index';
+import { InstallSource, ArchivePackage, RegistryPackage } from '../../../../common';
 
-const cache: Map<string, Buffer> = new Map();
-export const cacheGet = (key: string) => cache.get(key);
-export const cacheSet = (key: string, value: Buffer) => cache.set(key, value);
-export const cacheHas = (key: string) => cache.has(key);
-export const cacheClear = () => cache.clear();
-export const cacheDelete = (key: string) => cache.delete(key);
+const archiveEntryCache: Map<ArchiveEntry['path'], ArchiveEntry['buffer']> = new Map();
+export const getArchiveEntry = (key: string) => archiveEntryCache.get(key);
+export const setArchiveEntry = (key: string, value: Buffer) => archiveEntryCache.set(key, value);
+export const hasArchiveEntry = (key: string) => archiveEntryCache.has(key);
+export const clearArchiveEntries = () => archiveEntryCache.clear();
+export const deleteArchiveEntry = (key: string) => archiveEntryCache.delete(key);
 
-const archiveFilelistCache: Map<string, string[]> = new Map();
-export const getArchiveFilelist = (name: string, version: string) =>
-  archiveFilelistCache.get(pkgToPkgKey({ name, version }));
+export interface SharedKey {
+  name: string;
+  version: string;
+  installSource: InstallSource;
+}
+type SharedKeyString = string;
 
-export const setArchiveFilelist = (name: string, version: string, paths: string[]) =>
-  archiveFilelistCache.set(pkgToPkgKey({ name, version }), paths);
+type ArchiveFilelist = string[];
+const archiveFilelistCache: Map<SharedKeyString, ArchiveFilelist> = new Map();
+export const getArchiveFilelist = (keyArgs: SharedKey) =>
+  archiveFilelistCache.get(sharedKey(keyArgs));
 
-export const deleteArchiveFilelist = (name: string, version: string) =>
-  archiveFilelistCache.delete(pkgToPkgKey({ name, version }));
+export const setArchiveFilelist = (keyArgs: SharedKey, paths: string[]) =>
+  archiveFilelistCache.set(sharedKey(keyArgs), paths);
+
+export const deleteArchiveFilelist = (keyArgs: SharedKey) =>
+  archiveFilelistCache.delete(sharedKey(keyArgs));
+
+const packageInfoCache: Map<SharedKeyString, ArchivePackage | RegistryPackage> = new Map();
+const sharedKey = ({ name, version, installSource }: SharedKey) =>
+  `${name}-${version}-${installSource}`;
+
+export const getPackageInfo = (args: SharedKey) => {
```

Comment:
I wonder if we should name this something different as we already have a getPackageInfo https://github.com/elastic/kibana/pull/83384/files#diff-baf8294e3f474e4f2436b376ab8b1c1c8a611c0d0455ff88e78563c613deeb4dL106

Reply:
I think it's ok for the names to be the same since they're in different services. I also think one might replace/wrap the other eventually. I'll leave as-is and we can rename later if we want.